### PR TITLE
feat: Add service account impersonation credentials factory.

### PR DESCRIPTION
**Pull request by hessjcg**
Creates a new kind of CredentialFactory that can impersonate credentials. 

Part of #1168

### CODE REVIEW

File: core/src/main/java/com/google/cloud/sql/core/ServiceAccountImpersonatingCredentialFactory.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    //    targetPrincipal – the service account to impersonate
+    //    delegates – the chained list of delegates required to grant the final access_token.
+    //      If set, the sequence of identities must have "Service Account Token Creator"
+    //      capability granted to the preceding identity. For example, if set to
+    //      [serviceAccountB, serviceAccountC], the sourceCredential must have the Token Creator
+    //      role on serviceAccountB. serviceAccountB must have the Token Creator on
+    //      serviceAccountC. Finally, C must have Token Creator on target_principal.
+    //      If unset, sourceCredential must have that role on targetPrincipal.
+    List<String> reversedDelegates = Lists.reverse(delegates);
+    this.targetPrincipal = reversedDelegates.get(reversedDelegates.size() - 1);
```
Comment (enocom) [inline core/src/main/java/com/google/cloud/sql/core/ServiceAccountImpersonatingCredentialFactory.java:74]: Should this just be `reversedDelegates.get(0)`?

File: core/src/main/java/com/google/cloud/sql/core/ServiceAccountImpersonatingCredentialFactory.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+  public GoogleCredentials getCredentials() {
+    GoogleCredentials credentials = source.getCredentials();
+
+    credentials =
+        ImpersonatedCredentials.newBuilder()
+            .setSourceCredentials(credentials)
+            .setTargetPrincipal(targetPrincipal)
+            .setDelegates(this.delegates)
+            .setScopes(
+                Arrays.asList(SQLAdminScopes.SQLSERVICE_ADMIN, SQLAdminScopes.CLOUD_PLATFORM))
```
Comment (enocom) [inline core/src/main/java/com/google/cloud/sql/core/ServiceAccountImpersonatingCredentialFactory.java:94]: How does this work with Auto IAM Login where the sqlservice.login scope is required?

File: core/src/test/java/com/google/cloud/sql/core/ServiceAccountImpersonatingCredentialFactoryTest.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+    // Test that the CredentialsFactory.create() works.
+    Credentials impersonatedCredentials = newCredentials(impersonatedFactory);
+    assertThat(impersonatedCredentials).isInstanceOf(ImpersonatedCredentials.class);
+
+    // Test that CredentialsFatory.getCredentials() works.
+    assertThat(impersonatedFactory.getCredentials()).isInstanceOf(ImpersonatedCredentials.class);
+    ImpersonatedCredentials ic = (ImpersonatedCredentials) impersonatedFactory.getCredentials();
+    assertThat(ic.getAccount()).isEqualTo("first@serviceaccount.com");
+    assertThat(ic.getSourceCredentials()).isEqualTo(credentials);
```
Comment (enocom) [inline core/src/test/java/com/google/cloud/sql/core/ServiceAccountImpersonatingCredentialFactoryTest.java:54]: Can we add an assertion here that checks the delegates?

Also shall we add a test for when delegates is empty?

File: core/src/main/java/com/google/cloud/sql/core/ServiceAccountImpersonatingCredentialFactory.java (hunk 1 above)
Comment (hessjcg) [reply core/src/main/java/com/google/cloud/sql/core/ServiceAccountImpersonatingCredentialFactory.java:74]: No. 
- First, this method receives `delegates` in accordance with the `--impersonate-service-account` contract, so  that the target credential is the first element.
- Next we reverse the list so that it is in order according to the Java ImpersonatedCredentials contract. The target credential is now the last element in `reversedDelegates`. 
- Then set this`this.targetPrincipal` to the last element in `reversedDelegates`
- Finally we set `this.delegates` to the beginning of the list, excluding the last element.

File: core/src/main/java/com/google/cloud/sql/core/ServiceAccountImpersonatingCredentialFactory.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+  @Override
+  public GoogleCredentials getCredentials() {
+    GoogleCredentials credentials = source.getCredentials();
+
+    credentials =
+        ImpersonatedCredentials.newBuilder()
+            .setSourceCredentials(credentials)
+            .setTargetPrincipal(targetPrincipal)
+            .setDelegates(this.delegates)
```
[the remark this answers is not in the document]
Comment (hessjcg) [reply core/src/main/java/com/google/cloud/sql/core/ServiceAccountImpersonatingCredentialFactory.java:92]: Yes

File: core/src/main/java/com/google/cloud/sql/core/ServiceAccountImpersonatingCredentialFactory.java (hunk 2 above)
Comment (hessjcg) [reply core/src/main/java/com/google/cloud/sql/core/ServiceAccountImpersonatingCredentialFactory.java:94]: To create a valid ImpersonatedCredential, you must set the scopes, but we don't need to set the `sqladmin.login` scope right now.

If IAM is enabled, the DefaultAccessTokenSupplier will use this ImpersonatedCredential instance to create downscoped credentials with a `sqladmin.login` scope.

File: core/src/test/java/com/google/cloud/sql/core/ServiceAccountImpersonatingCredentialFactoryTest.java (hunk 3 above)
Comment (hessjcg) [reply core/src/test/java/com/google/cloud/sql/core/ServiceAccountImpersonatingCredentialFactoryTest.java:76]: Unfortunately, we can't test delegates. ImpersonatedCredentials does not expose delegates in its public api.

I'll add a test case where `delegates` contains a single element.

File: core/src/main/java/com/google/cloud/sql/core/ServiceAccountImpersonatingCredentialFactory.java (hunk 1 above)
Comment (enocom) [reply core/src/main/java/com/google/cloud/sql/core/ServiceAccountImpersonatingCredentialFactory.java:74]: Sorry I misspoke. I mean, should this just be `delegates.get(0)`?

Comment (hessjcg) [reply core/src/main/java/com/google/cloud/sql/core/ServiceAccountImpersonatingCredentialFactory.java:74]: Well, that would return the same thing. But I think it is more straightforward to only use reversedDelegates.

**Review by enocom — APPROVED** [review]
